wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/convert/DefaultEditorProvisioner.java

```diff
@@ -99,20 +104,97 @@ public void apply(DevfileImpl devfile, FileContentProvider contentProvider)
+      Map<String, String> missingPluginsIdToRef)
+      throws DevfileException {
     for (ComponentImpl t : components) {
       if (PLUGIN_COMPONENT_TYPE.equals(t.getType())) {
         String pluginPublisherAndName = getPluginPublisherAndName(t, contentProvider);
         missingPluginsIdToRef.remove(pluginPublisherAndName);
       }
     }
+  }
 
-    missingPluginsIdToRef
-        .values()
-        .forEach(pluginRef -> components.add(new ComponentImpl(PLUGIN_COMPONENT_TYPE, pluginRef)));
+  /**
+   * Tries to add default plugins to the Devfile components. Each plugin is initially parsed by
+   * plugin ref. If the plugin does not have a reference, it is added to the component list, and its
+   * plugin ID will be used to resolve it. If it has a reference, the Plugin is evaluated, so that
+   * its meta.yaml can be retrieved. From the meta.yaml, the new Component's ID and reference are
+   * properly set, and the Component is added to the list.
+   *
+   * @param components - The list of Devfile components
+   * @param contentProvider - The content provider to retrieve YAML
+   * @param missingPluginsIdToRef - The list of default plugins that are not currently in the list
+   *     of devfile components
+   * @throws DevfileException
+   */
+  private void addMissingPlugins(
+      List<ComponentImpl> components,
+      FileContentProvider contentProvider,
+      Map<String, String> missingPluginsIdToRef)
+      throws DevfileException {
+    for (String pluginRef : missingPluginsIdToRef.values()) {
+      ComponentImpl component;
+      try {
+        ExtendedPluginFQN fqn = pluginFQNParser.parsePluginFQN(pluginRef);
+        if (!isNullOrEmpty(fqn.getId())) {
+          component = new ComponentImpl(PLUGIN_COMPONENT_TYPE, pluginRef);
+        } else {
+          component = createReferencePluginComponent(pluginRef, contentProvider);
+        }
+        components.add(component);
+      } catch (InfrastructureException e) {
+        throw new DevfileException(e.getMessage(), e);
+      }
+    }
+  }
+
+  /**
+   * Evaluates a plugin FQN by retrieving it's meta.yaml, and sets it's name and reference to the
+   * appropriate values.
+   *
+   * @param pluginRef - The plugin ref string
+   * @param contentProvider - The content provider used to read YAML data
+   * @return - A {@link ComponentImpl} with it's ID and reference URL set.
+   * @throws InfrastructureException
+   */
+  private ComponentImpl createReferencePluginComponent(
+      String pluginRef, FileContentProvider contentProvider) throws InfrastructureException {
+    ExtendedPluginFQN fqn = pluginFQNParser.evaluateFqn(pluginRef, contentProvider);
+    ComponentImpl component = new ComponentImpl();
+    component.setType(PLUGIN_COMPONENT_TYPE);
+    if (!isNullOrEmpty(fqn.getId())) {
+      component.setId(fqn.getId());
+    }
+    if (!isNullOrEmpty(fqn.getReference())) {
+      component.setReference(fqn.getReference());
+    }
+    return component;
```

Comment:
is it ok if I end here with only one of `id` and `reference` inside new component?

Reply:
Yes, this was how it worked before the refactor.  e.g. the default plugin `eclipse/che-machine-exec-plugin/nightly` would only have an ID, and no reference.  A Component created from a URL would have an `id` of null and a set `reference`.

Reply:
Technically speaking, this code _should_ (as far as I understand) always have reference and id set, since `evaluateFqn` fetches the meta.yaml and throws an `InfrastructureException` if any id field is empty. Similarly, `fqn.getReference()` is just set to `pluginRef` from earlier. However, the javadoc for that method doesn't guarantee this, so it maybe makes sense to null check anyways.